File: packages/mobile/native/android/java/com/bendyline/gezel/mobile/ModelDownloads.java
package com.bendyline.gezel.mobile;

import android.util.AtomicFile;
import com.getcapacitor.JSObject;
import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.security.MessageDigest;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;
import java.util.function.BooleanSupplier;
import org.json.*;

/** Foreground, resumable verified acquisition. Never publishes partial weights. */
final class ModelDownloads {
    interface Connection extends AutoCloseable {
        int status() throws Exception;
        String header(String name);
        InputStream body() throws Exception;
        void close();
    }
    interface Transport {
        Connection open(URI url, String method, Map<String,String> headers) throws Exception;
        default Connection open(URI url,String method,Map<String,String> headers,BooleanSupplier cancelled) throws Exception {
            if(cancelled.getAsBoolean())throw new IOException("Download paused");
            return open(url,method,headers);
        }
        default void cancel(Thread thread) {}
    }
    interface SourceResult { void complete(JSONObject source, Exception error); }
    private static final class HttpTransport implements Transport {
        private final ConcurrentHashMap<Thread,HttpURLConnection> pending = new ConcurrentHashMap<>();
        public void cancel(Thread thread) { if (thread != null) { HttpURLConnection value=pending.get(thread); if(value!=null)value.disconnect(); } }
        public Connection open(URI initial,String method,Map<String,String> headers)throws Exception {return open(initial,method,headers,()->false);}
        public Connection open(URI initial, String method, Map<String,String> headers,BooleanSupplier cancelled) throws Exception {
            URI url = initial;
            for (int redirects=0; redirects<6; redirects++) {
                ModelDownloadSource.validateURL(url);
                HttpURLConnection connection = (HttpURLConnection) url.toURL().openConnection();
                connection.setInstanceFollowRedirects(false);
                connection.setRequestMethod(method);
                connection.setConnectTimeout(20_000); connection.setReadTimeout(30_000);
                connection.setRequestProperty("Accept-Encoding", "identity");
                for (Map.Entry<String,String> header : headers.entrySet()) connection.setRequestProperty(header.getKey(), header.getValue());
                Thread owner=Thread.currentThread(); pending.put(owner,connection);
                int status;
                try { if(cancelled.getAsBoolean())throw new IOException("Download paused"); status = connection.getResponseCode(); }
                catch (Exception error) { pending.remove(owner,connection); connection.disconnect(); throw error; }
                if (status == 301 || status == 302 || status == 303 || status == 307 || status == 308) {
                    String location = connection.getHeaderField("Location"); pending.remove(owner,connection); connection.disconnect();
                    if (location == null) throw new IOException("Model source omitted its redirect location");
                    url = url.resolve(location); continue;
                }
                return new Connection() {
                    public int status() { return status; }
                    public String header(String name) { return connection.getHeaderField(name); }
                    public InputStream body() throws Exception { return connection.getInputStream(); }
                    public void close() { pending.remove(owner,connection); connection.disconnect(); }
                };
            }
            throw new IOException("Model source redirected too many times");
        }
    }
    private final MobileStore store;
    private final File root;
    private final Transport transport;
    private final ExecutorService worker = Executors.newSingleThreadExecutor();
    private final ExecutorService metadataWorker = Executors.newSingleThreadExecutor();
    private volatile Thread downloadThread, sourceThread;
    private volatile boolean sourceCancelled;
    private boolean sourcePending;
    private String activeId;
    private volatile boolean cancelled;
    private volatile Connection connection;
    private Future<?> active;

    ModelDownloads(MobileStore store) throws Exception { this(store, new HttpTransport()); }
    ModelDownloads(MobileStore store, Transport transport) throws Exception {
        this.store = store; this.root = store.downloadsRoot(); this.transport = transport;
        for (JSONObject record : records()) {
            String state = record.getString("state");
            if (state.equals("queued") || state.equals("downloading") || state.equals("verifying")) {
                record.put("state", "paused").put("error", "Download paused when the app closed. Resume when ready.");
                record.put("downloadedBytes", candidate(record).isFile() ? candidate(record).length() : 0);
                save(record);
            }
        }
    }
    private static String id(String value) throws IOException {
        try { if (UUID.fromString(value).toString().equals(value)) return value; } catch (Exception ignored) {}
        throw new IOException("Invalid download identity");
    }
    private File folder(String value) throws Exception {
        File folder = new File(root, id(value));
        if (Files.isSymbolicLink(folder.toPath()) || !folder.getCanonicalFile().equals(folder.getAbsoluteFile())) throw new IOException("Invalid download storage");
        return folder;
    }
    private File partial(JSONObject record) throws Exception {
        File file=new File(folder(record.getString("id")),"model.part");
        if(Files.isSymbolicLink(file.toPath())||!file.getCanonicalFile().equals(file.getAbsoluteFile()))throw new IOException("Invalid model staging file");
        return file;
    }
    private File candidate(JSONObject record) throws Exception {
        File part = partial(record);
        return part.isFile() ? part : store.downloadModelPath(record.getString("id"));
    }
    private JSONObject read(String value) throws Exception {
        File path = new File(folder(value), "download.json");
        if (Files.isSymbolicLink(path.toPath())) throw new IOException("Invalid download record");
        JSONObject record;
        try (FileInputStream file = new AtomicFile(path).openRead()) {
            long size=file.getChannel().size(); if(size>32*1024)throw new IOException("Download record exceeds its limit");
            byte[] bytes=new byte[(int)size]; new DataInputStream(file).readFully(bytes);
            record=new JSONObject(new String(bytes,StandardCharsets.UTF_8));
        }
        if (!value.equals(record.getString("id"))) throw new IOException("Download identity changed");
        record.put("source", ModelDownloadSource.validate(record.getJSONObject("source"), true));
        if (!Set.of("queued","downloading","paused","verifying","complete","failed").contains(record.getString("state"))) throw new IOException("Invalid download state");
        if (!record.getString("state").equals("complete")) {
            File file=candidate(record);
            record.put("downloadedBytes",Math.min(file.isFile()?file.length():0,record.getJSONObject("source").getLong("sizeBytes")));
        }
        return record;
    }
    private List<JSONObject> records() throws Exception {
        List<JSONObject> result = new ArrayList<>();
        File[] paths = root.listFiles();
        if (paths == null) throw new IOException("Download storage is unavailable");
        for (File path : paths) if (path.isDirectory()) { result.add(read(path.getName())); if (result.size()>16) throw new IOException("Download library exceeds its limit"); }
        return result;
    }
    private void save(JSONObject record) throws Exception {
        File directory = folder(record.getString("id"));
        if (!directory.isDirectory() && !directory.mkdirs()) throw new IOException("Cannot save download state");
        AtomicFile file = new AtomicFile(new File(directory, "download.json"));
        FileOutputStream output = null;
        try { output = file.startWrite(); output.write(record.toString().getBytes(StandardCharsets.UTF_8)); file.finishWrite(output); }
        catch (Exception error) { file.failWrite(output); throw error; }
    }
    private JSObject publicRecord(JSONObject record) throws Exception {
        JSObject result = new JSObject().put("id",record.getString("id")).put("name",record.getString("name"))
            .put("source",record.getJSONObject("source")).put("state",record.getString("state")).put("downloadedBytes",record.getLong("downloadedBytes"));
        if (record.has("error")) result.put("error",record.getString("error"));
        if (record.has("modelId")) result.put("modelId",record.getString("modelId"));
        return result;
    }
    synchronized JSObject list() throws Exception {
        JSONArray downloads = new JSONArray();
        for (JSONObject record : records()) downloads.put(publicRecord(record));
        return new JSObject().put("downloads",downloads);
    }
    synchronized void resolveSource(JSONObject raw, SourceResult completion) throws Exception {
        if(sourcePending)throw new IOException("A model source is already being checked");
        JSONObject source=ModelDownloadSource.validate(raw,false);
        sourcePending=true;sourceCancelled=false;
        metadataWorker.execute(()->{
            JSONObject result=null;Exception failure=null;sourceThread=Thread.currentThread();
            try {
                if(sourceCancelled)throw new IOException("Source lookup cancelled");
                try(Connection response=transport.open(ModelDownloadSource.url(source),"HEAD",Map.of(),()->sourceCancelled)) {
                    if(response.status()!=200)throw new IOException("Model source cannot be inspected (HTTP "+response.status()+"). Gated models must be imported from Files.");
                    long size=parseLength(response.header("Content-Length"));
                    if(size<4||size>MobileStore.MAX_MODEL)throw new IOException("This model exceeds the 4 GiB mobile download limit");
                    result=source.put("sizeBytes",size);
                }
            }catch(Exception error){failure=error;}
            synchronized(this){
                if(sourceCancelled)failure=new IOException("Source lookup cancelled");
                sourceThread=null;sourcePending=false;
            }
            completion.complete(failure==null?result:null,failure);
        });
    }
    void cancelSourceResolution() { sourceCancelled=true;transport.cancel(sourceThread); }
    synchronized JSObject start(JSONObject raw, String name) throws Exception {
        if (activeId != null) throw new IOException("A model download is already running");
        JSONObject source=ModelDownloadSource.validate(raw,true);
        if (name==null || name.isEmpty() || name.length()>200 || name.indexOf(0)>=0) throw new IOException("Invalid model name");
        if (records().size()>=16) throw new IOException("Remove a saved download before starting another");
        if (root.getUsableSpace()<source.getLong("sizeBytes")+64L*1024*1024) throw new IOException("Not enough storage for this model");
        JSONObject record=new JSONObject().put("id",UUID.randomUUID().toString()).put("name",name).put("source",source).put("state","queued").put("downloadedBytes",0);
        save(record); launch(record.getString("id"));
        return new JSObject().put("download",publicRecord(record));
    }
    synchronized JSObject resume(String value) throws Exception {
        if (activeId != null) throw new IOException("A model download is already running");
        JSONObject record=read(value);
        if (record.getString("state").equals("complete")) return new JSObject().put("download",publicRecord(record));
        long remaining=record.getJSONObject("source").getLong("sizeBytes")-(candidate(record).isFile()?candidate(record).length():0);
        if (root.getUsableSpace()<remaining+64L*1024*1024) throw new IOException("Not enough storage to resume this download");
        record.put("state","queued"); record.remove("error"); save(record); launch(value);
        return new JSObject().put("download",publicRecord(record));
    }
    private void launch(String value) {
        activeId=value; cancelled=false;
        active=worker.submit(()->transfer(value));
    }
    void cancel(String value) throws Exception {
        Future<?> pending;
        synchronized(this) { id(value); if (!value.equals(activeId)) return; cancelled=true; pending=active; }
        transport.cancel(downloadThread); Connection current=connection; if (current!=null) current.close();
        if (pending!=null) pending.get(35,TimeUnit.SECONDS);
    }
    void requestPause() { cancelled=true;sourceCancelled=true; }
    void suspend() { requestPause();cancelSourceResolution();transport.cancel(downloadThread);Connection current=connection;if(current!=null)current.close(); }
    void close() { suspend(); worker.shutdown(); metadataWorker.shutdown(); }
    void remove(String value) throws Exception {
        cancel(value);
        synchronized(this) {
            read(value);
            File directory=folder(value);
            for(File file : Objects.requireNonNull(directory.listFiles())) Files.delete(file.toPath());
            Files.delete(directory.toPath());
        }
    }
    private static long parseLength(String raw) throws IOException {
        if(raw==null || !raw.matches("[0-9]{1,12}"))throw new IOException("Model source omitted a valid exact length");
        try{return Long.parseLong(raw);}catch(NumberFormatException error){throw new IOException("Invalid source length");}
    }
    private static String etag(Connection response) {
        String value=response.header("ETag");
        return value!=null && value.length()<=512 && value.matches("\"[^\\r\\n\"]+\"")?value:null;
    }
    private void checkpoint(JSONObject record,String state,long bytes) throws Exception {
        synchronized(this) {record.put("state",state).put("downloadedBytes",bytes);save(record);}
    }
    private void checkCancelled() throws IOException { if(cancelled)throw new IOException("Download paused"); }
    private void transfer(String value) {
        JSONObject record=null;downloadThread=Thread.currentThread();
        try {
            synchronized(this) {record=read(value);}
            JSONObject source=record.getJSONObject("source");long expected=source.getLong("sizeBytes");
            File part=partial(record), candidate=candidate(record);long offset=candidate.isFile()?candidate.length():0;
            if(offset>expected)throw new IOException("Saved partial exceeds expected model length");
            checkCancelled();
            if(offset<expected) {
                // Without a strong validator the old bytes cannot be resumed.
                String prior=record.optString("etag",null);
                if(offset>0 && prior==null) {Files.deleteIfExists(part.toPath());offset=0;}
                Map<String,String> headers=offset>0?Map.of("Range","bytes="+offset+"-","If-Range",prior):Map.of();
                Connection response=transport.open(ModelDownloadSource.url(source),"GET",headers,()->cancelled);connection=response;
                try(response) {
                    checkCancelled();int status=response.status();String current=etag(response);
                    if(status==200) {offset=0;}
                    else if(status==206 && offset>0) {
                        Matcher range=Pattern.compile("bytes ([0-9]+)-([0-9]+)/([0-9]+)").matcher(Objects.toString(response.header("Content-Range"),""));
                        if(current==null || !current.equals(prior) || !range.matches() || Long.parseLong(range.group(1))!=offset || Long.parseLong(range.group(2))!=expected-1 || Long.parseLong(range.group(3))!=expected)
                            throw new IOException("Model source changed its resume identity or byte range; remove this download and try again");
                    } else throw new IOException("Model download failed (HTTP "+status+")");
                    if(parseLength(response.header("Content-Length"))!=expected-offset)throw new IOException("Model source length differs from the verified download source");
                    String encoding=response.header("Content-Encoding");if(encoding!=null && !encoding.equalsIgnoreCase("identity"))throw new IOException("Compressed model transfers are not supported");
                    if(current==null)record.remove("etag");else record.put("etag",current);
                    try(FileOutputStream output=new FileOutputStream(part,offset>0);InputStream input=response.body()) {
                        checkpoint(record,"downloading",offset);long last=offset;byte[] buffer=new byte[64*1024];int count;
                        while((count=input.read(buffer))!=-1) {
                            checkCancelled();if(offset+count>expected)throw new IOException("Model transfer exceeded its expected length");
                            output.write(buffer,0,count);offset+=count;
                            if(offset-last>=1024*1024) {output.getFD().sync();checkpoint(record,"downloading",offset);last=offset;}
                        }
                        output.getFD().sync();checkpoint(record,"verifying",offset);
                    }
                    if(offset!=expected)throw new IOException("Model transfer ended early. Resume to continue.");
                } finally {connection=null;}
                candidate=part;
            }
            checkCancelled();checkpoint(record,"verifying",expected);
            MessageDigest digest=MessageDigest.getInstance("SHA-256");
            try(InputStream input=new FileInputStream(candidate)) {byte[] buffer=new byte[1024*1024];int count;while((count=input.read(buffer))!=-1){checkCancelled();digest.update(buffer,0,count);}}
            StringBuilder hash=new StringBuilder();for(byte b:digest.digest())hash.append(String.format(Locale.ROOT,"%02x",b));
            if(!hash.toString().equals(source.getString("sha256"))) {Files.deleteIfExists(candidate.toPath());record.remove("etag");record.put("downloadedBytes",0);throw new IOException("Model SHA-256 verification failed. No model was installed.");}
            checkCancelled();
            synchronized(this) {
                checkCancelled();store.publishDownloadedModel(value,record.getString("name"),source,candidate);
                record.put("modelId",value);record.remove("error");checkpoint(record,"complete",expected);
            }
        } catch(Exception failure) {
            if(record!=null)try {
                File saved=candidate(record);long bytes=saved.isFile()?saved.length():0;
                String message=cancelled?"Download paused. Resume when ready.":Objects.toString(failure.getMessage(),"Model download failed");
                record.put("error",message.substring(0,Math.min(message.length(),1000)));
                checkpoint(record,cancelled?"paused":"failed",Math.min(bytes,record.getJSONObject("source").getLong("sizeBytes")));
            }catch(Exception ignored) { /* Existing durable checkpoint stays resumable after storage recovers. */ }
        } finally {connection=null;downloadThread=null;synchronized(this){activeId=null;active=null;}}
    }
}

File: packages/mobile/native/ios/Sources/GezelMobileStorage/ModelDownloads.swift
import Foundation
import CryptoKit

public struct MobileModelDownload: Codable, Sendable {
    public let id: String
    public let name: String
    public let source: MobileModelSource
    public var state: String
    public var downloadedBytes: Int64
    public var error: String?
    public var modelId: String?
    var etag: String?
    public func json() throws -> [String: Any] {
        var result = try JSONSerialization.jsonObject(with: JSONEncoder().encode(self)) as! [String:Any]
        result.removeValue(forKey: "etag")
        return result
    }
}

/// Foreground streaming acquisition, independent of inference and product files.
public final class ModelDownloads: @unchecked Sendable {
    private let store: MobileStore
    private let root: URL
    private let configuration: URLSessionConfiguration
    private let queue = DispatchQueue(label: "com.bendyline.gezel.model-download")
    private let sourceQueue = DispatchQueue(label: "com.bendyline.gezel.model-source")
    private let journalLock = NSRecursiveLock()
    private var sourcePending = false
    private var sourceCancelled = false
    private var sourceTransfer: ModelDownloadHTTP?
    private let cancellationLock = NSLock()
    private var cancelled = false
    private var cancelIdentity: String?
    private var activeId: String?
    private var transfer: ModelDownloadHTTP?
    private var output: FileHandle?
    private var current: MobileModelDownload?
    private var offset: Int64 = 0
    private var checkpoint: Int64 = 0
    private var cancelCallbacks: [@Sendable () -> Void] = []
    private let fm = FileManager.default

    public init(store: MobileStore, configuration: URLSessionConfiguration = .ephemeral) throws {
        self.store = store; self.root = try store.downloadsDirectory(); self.configuration = configuration
        for var record in try records() where ["queued","downloading","verifying"].contains(record.state) {
            record.state = "paused"; record.error = "Download paused when the app closed. Resume when ready."
            record.downloadedBytes = try length(candidate(record))
            try save(record)
        }
    }
    private func folder(_ id: String) throws -> URL {
        guard UUID(uuidString: id)?.uuidString.lowercased() == id else { throw ModelDownloadError("Invalid download identity") }
        let url = root.appendingPathComponent(id, isDirectory: true)
        try rejectLink(url)
        return url
    }
    private func rejectLink(_ file: URL) throws {
        // attributesOfItem inspects dangling links too; fileExists follows them.
        if let attributes = try? fm.attributesOfItem(atPath: file.path), attributes[.type] as? FileAttributeType == .typeSymbolicLink {
            throw ModelDownloadError("Invalid model staging link")
        }
    }
    private func partial(_ record: MobileModelDownload) throws -> URL {
        let file = try folder(record.id).appendingPathComponent("model.part")
        try rejectLink(file); return file
    }
    private func candidate(_ record: MobileModelDownload) throws -> URL {
        let file = try partial(record)
        return fm.fileExists(atPath: file.path) ? file : try store.downloadModelURL(id: record.id)
    }
    private func length(_ file: URL) throws -> Int64 {
        guard fm.fileExists(atPath: file.path) else { return 0 }
        let values = try file.resourceValues(forKeys: [.isRegularFileKey,.isSymbolicLinkKey,.fileSizeKey])
        guard values.isRegularFile == true, values.isSymbolicLink != true else { throw ModelDownloadError("Invalid model staging file") }
        return Int64(values.fileSize ?? 0)
    }
    private func read(_ id: String) throws -> MobileModelDownload {
        let url = try folder(id).appendingPathComponent("download.json")
        try rejectLink(url)
        guard try length(url) <= 32 * 1024 else { throw ModelDownloadError("Invalid download record") }
        var record = try JSONDecoder().decode(MobileModelDownload.self, from: Data(contentsOf: url))
        guard record.id == id, !record.name.isEmpty, record.name.utf16.count <= 200,
              ["queued","downloading","paused","verifying","complete","failed"].contains(record.state) else { throw ModelDownloadError("Invalid download record") }
        try record.source.validate(exact: true)
        if record.state != "complete" { record.downloadedBytes = min(try length(candidate(record)), record.source.sizeBytes!) }
        return record
    }
    private func records() throws -> [MobileModelDownload] {
        journalLock.lock(); defer { journalLock.unlock() }
        let entries = try fm.contentsOfDirectory(at: root, includingPropertiesForKeys: [.isDirectoryKey])
        guard entries.count <= 16 else { throw ModelDownloadError("Download library exceeds its limit") }
        return try entries.map { try read($0.lastPathComponent) }
    }
    private func save(_ record: MobileModelDownload) throws {
        journalLock.lock(); defer { journalLock.unlock() }
        let directory = try folder(record.id)
        try fm.createDirectory(at: directory, withIntermediateDirectories: true)
        try JSONEncoder().encode(record).write(to: directory.appendingPathComponent("download.json"), options: .atomic)
    }
    private func checkCancelled() throws {
        cancellationLock.lock(); let stopped = cancelled; cancellationLock.unlock()
        if stopped { throw ModelDownloadError("Download paused") }
    }
    public func list() throws -> [MobileModelDownload] { try records() }
    private static func exactLength(_ text: String?) throws -> Int64 {
        guard let text, text.range(of: "^[0-9]{1,12}$", options: .regularExpression) != nil, let value = Int64(text) else {
            throw ModelDownloadError("Model source omitted a valid exact length")
        }
        return value
    }
    private static func etag(_ response: HTTPURLResponse) -> String? {
        guard let value = response.value(forHTTPHeaderField: "ETag"), value.utf8.count <= 512,
              value.range(of: "^\"[^\\r\\n\"]+\"$", options: .regularExpression) != nil else { return nil }
        return value
    }
    public func resolve(_ identity: MobileModelSource, completion: @escaping @Sendable (Result<MobileModelSource,Error>) -> Void) {
        cancellationLock.lock()
        guard !sourcePending else { cancellationLock.unlock(); completion(.failure(ModelDownloadError("A model source is already being checked"))); return }
        sourcePending = true; sourceCancelled = false; cancellationLock.unlock()
        sourceQueue.async {
            let done: @Sendable (Result<MobileModelSource,Error>) -> Void = { result in
                self.cancellationLock.lock()
                let cancelled = self.sourceCancelled
                self.sourceTransfer = nil; self.sourcePending = false; self.cancellationLock.unlock()
                completion(cancelled ? .failure(ModelDownloadError("Source lookup cancelled")) : result)
            }
            do {
                try identity.validate(exact: false)
                let box = ResolvedSource(identity)
                let request = try ModelDownloadHTTP(url: identity.url(), method: "HEAD", fields: [:], configuration: self.configuration,
                    queue: self.sourceQueue, headers: { response in
                        guard response.statusCode == 200 else { throw ModelDownloadError("Model source cannot be inspected (HTTP \(response.statusCode)). Gated models must be imported from Files.") }
                        box.source.sizeBytes = try Self.exactLength(response.value(forHTTPHeaderField: "Content-Length"))
                        try box.source.validate(exact: true)
                    }, chunk: { _ in throw ModelDownloadError("Model metadata request returned unexpected bytes") }, completion: { error in
                        if let error { done(.failure(error)) } else { done(.success(box.source)) }
                    })
                self.cancellationLock.lock()
                self.sourceTransfer = request
                let cancelled = self.sourceCancelled
                self.cancellationLock.unlock()
                request.start(); if cancelled { request.cancel() }
            } catch { done(.failure(error)) }
        }
    }
    public func cancelSourceResolution() {
        cancellationLock.lock(); sourceCancelled = true; let request = sourceTransfer; cancellationLock.unlock()
        request?.cancel()
    }
    public func suspend() { cancelSourceResolution(); cancel() }
    private final class ResolvedSource: @unchecked Sendable {
        var source: MobileModelSource
        init(_ source: MobileModelSource) { self.source = source }
    }
    private func space(for remaining: Int64) throws {
        let free = try fm.attributesOfFileSystem(forPath: root.path)[.systemFreeSize] as? NSNumber
        guard (free?.int64Value ?? 0) > remaining + 64 * 1024 * 1024 else { throw ModelDownloadError("Not enough storage for this model") }
    }
    public func start(source: MobileModelSource, name: String) throws -> MobileModelDownload {
        try queue.sync {
            guard activeId == nil else { throw ModelDownloadError("A model download is already running") }
            try source.validate(exact: true)
            guard !name.isEmpty, name.utf16.count <= 200, !name.contains("\0") else { throw ModelDownloadError("Invalid model name") }
            guard try records().count < 16 else { throw ModelDownloadError("Remove a saved download before starting another") }
            try space(for: source.sizeBytes!)
            let record = MobileModelDownload(id: UUID().uuidString.lowercased(), name: name, source: source, state: "queued", downloadedBytes: 0)
            try save(record); launch(record)
            return record
        }
    }
    public func resume(id: String) throws -> MobileModelDownload {
        try queue.sync {
            guard activeId == nil else { throw ModelDownloadError("A model download is already running") }
            var record = try read(id)
            if record.state == "complete" { return record }
            try space(for: record.source.sizeBytes! - record.downloadedBytes)
            record.state = "queued"; record.error = nil; try save(record); launch(record)
            return record
        }
    }
    private func launch(_ record: MobileModelDownload) {
        cancellationLock.lock(); cancelled = false; cancelIdentity = record.id; cancellationLock.unlock()
        activeId = record.id
        queue.async { self.begin(record) }
    }
    public func cancel(id: String? = nil, completion: @escaping @Sendable () -> Void = {}) {
        // Interrupt hashing too; it deliberately checks between bounded chunks.
        cancellationLock.lock()
        guard cancelIdentity != nil, id == nil || cancelIdentity == id else { cancellationLock.unlock(); completion(); return }
        cancelled = true; cancellationLock.unlock()
        queue.async {
            if let id, self.activeId != id { completion(); return }
            guard self.activeId != nil else { completion(); return }
            self.cancelCallbacks.append(completion)
            if let transfer = self.transfer { transfer.cancel() }
            else { self.finish(ModelDownloadError("Download paused")) }
        }
    }
    public func remove(id: String, completion: @escaping @Sendable (Result<Void,Error>) -> Void) {
        cancel(id: id) {
            self.queue.async {
                do { _ = try self.read(id); try self.fm.removeItem(at: self.folder(id)); completion(.success(())) }
                catch { completion(.failure(error)) }
            }
        }
    }
    private func begin(_ input: MobileModelDownload) {
        current = input
        do {
            let record = input; let expected = record.source.sizeBytes!
            let file = try candidate(record); offset = try length(file)
            guard offset <= expected else { throw ModelDownloadError("Saved partial exceeds expected model length") }
            try checkCancelled()
            if offset == expected { try verify(file); return }
            let part = try partial(record)
            if offset > 0 && record.etag == nil { try fm.removeItem(at: part); offset = 0 }
            let priorOffset = offset, priorTag = record.etag
            let fields = offset > 0 ? ["Range":"bytes=\(offset)-", "If-Range":record.etag!] : [:]
            transfer = try ModelDownloadHTTP(url: record.source.url(), method: "GET", fields: fields, configuration: configuration, queue: queue,
                headers: { response in
                    try self.checkCancelled()
                    let tag = Self.etag(response)
                    if response.statusCode == 200 { self.offset = 0 }
                    else if response.statusCode == 206, priorOffset > 0 {
                        let expectedRange = "bytes \(priorOffset)-\(expected-1)/\(expected)"
                        guard let tag, tag == priorTag, response.value(forHTTPHeaderField: "Content-Range") == expectedRange else {
                            throw ModelDownloadError("Model source changed its resume identity or byte range; remove this download and try again")
                        }
                    } else { throw ModelDownloadError("Model download failed (HTTP \(response.statusCode))") }
                    guard try Self.exactLength(response.value(forHTTPHeaderField: "Content-Length")) == expected - self.offset else { throw ModelDownloadError("Model source length differs from the verified download source") }
                    if let encoding = response.value(forHTTPHeaderField: "Content-Encoding"), encoding.lowercased() != "identity" { throw ModelDownloadError("Compressed model transfers are not supported") }
                    if !self.fm.fileExists(atPath: part.path) { guard self.fm.createFile(atPath: part.path, contents: nil) else { throw ModelDownloadError("Cannot create model staging file") } }
                    self.output = try FileHandle(forWritingTo: part)
                    if self.offset == 0 { try self.output!.truncate(atOffset: 0) }
                    else { guard try self.output!.seekToEnd() == UInt64(self.offset) else { throw ModelDownloadError("Saved partial changed") } }
                    self.current!.etag = tag; self.checkpoint = self.offset
                    try self.update(state: "downloading")
                }, chunk: { data in
                    try self.checkCancelled()
                    guard self.offset + Int64(data.count) <= expected else { throw ModelDownloadError("Model transfer exceeded its expected length") }
                    try self.output!.write(contentsOf: data); self.offset += Int64(data.count)
                    if self.offset - self.checkpoint >= 1024 * 1024 {
                        try self.output!.synchronize(); try self.update(state: "downloading"); self.checkpoint = self.offset
                    }
                }, completion: { error in
                    do {
                        try self.output?.synchronize(); try self.output?.close(); self.output = nil; self.transfer = nil
                        if let error { throw error }
                        guard self.offset == expected else { throw ModelDownloadError("Model transfer ended early. Resume to continue.") }
                        try self.verify(part)
                    } catch { self.finish(error) }
                })
            transfer!.start()
        } catch { finish(error) }
    }
    private func update(state: String) throws {
        current!.state = state; current!.downloadedBytes = offset
        try save(current!)
    }
    private func verify(_ file: URL) throws {
        try checkCancelled(); try update(state: "verifying")
        let input = try FileHandle(forReadingFrom: file); defer { try? input.close() }
        var digest = SHA256()
        while let chunk = try input.read(upToCount: 1024 * 1024), !chunk.isEmpty { try checkCancelled(); digest.update(data: chunk) }
        let hash = digest.finalize().map { String(format: "%02x", $0) }.joined()
        guard hash == current!.source.sha256 else {
            try fm.removeItem(at: file); offset = 0; current!.etag = nil
            throw ModelDownloadError("Model SHA-256 verification failed. No model was installed.")
        }
        try checkCancelled()
        _ = try store.publishDownloadedModel(id: current!.id, name: current!.name, source: current!.source, file: file)
        current!.modelId = current!.id; current!.error = nil; try update(state: "complete")
        finish(nil)
    }
    private func finish(_ error: Error?) {
        try? output?.synchronize(); try? output?.close(); output = nil; transfer = nil
        if let error, var record = current {
            cancellationLock.lock(); let paused = cancelled; cancellationLock.unlock()
            record.state = paused ? "paused" : "failed"
            record.error = String((paused ? "Download paused. Resume when ready." : error.localizedDescription).prefix(1000))
            record.downloadedBytes = min((try? length(candidate(record))) ?? 0, record.source.sizeBytes!)
            try? save(record)
        }
        current = nil; activeId = nil
        cancellationLock.lock(); cancelIdentity = nil; cancellationLock.unlock()
        let callbacks = cancelCallbacks; cancelCallbacks.removeAll()
        callbacks.forEach { $0() }
    }
}
